src/lib.rs

```diff
@@ -2,14 +2,165 @@ use zed_extension_api::{
     self as zed, lsp::CompletionKind, settings::LspSettings, CodeLabel, CodeLabelSpan,
 };
 
-struct Java;
+struct Java {
+    cached_binary_path: Option<String>,
+}
+
+impl Java {
+    fn language_server_binary_path(
+        &mut self,
+        language_server_id: &zed::LanguageServerId,
+        worktree: &zed::Worktree,
+    ) -> zed::Result<String> {
+        // Quickly return if the binary path is already cached
+        // Expect binary path to be validated when setting the cache so no checking is done here
+        if let Some(path) = &self.cached_binary_path {
+            return Ok(path.clone());
+        }
+
+        // Determine the binary name based on the current platform
+        let (platform, _) = zed::current_platform();
+        let binary_name = match platform {
+            zed::Os::Windows => "jdtls.bat",
+            _ => "jdtls",
+        }
+        .to_string();
+
+        // Use binary available on PATH if it exists
+        if let Some(path) = worktree.which(&binary_name) {
+            // Probably we want to check if the binary is executable too here
+            if std::fs::metadata(&path).map_or(false, |stat| stat.is_file()) {
+                self.cached_binary_path = Some(path.clone());
+                return Ok(path.clone());
+            }
+        }
+
+        // Attempt to install locally
+
+        zed::set_language_server_installation_status(
+            language_server_id,
+            &zed::LanguageServerInstallationStatus::CheckingForUpdate,
+        );
+
+        // Use version specified in settings or latest version released on Github
+        let version = match LspSettings::for_worktree(language_server_id.as_ref(), worktree)?
+            .settings
+            .and_then(|settings| {
+                settings.get("version").and_then(|version_value| {
+                    version_value
+                        .as_str()
+                        .map(|version_str| version_str.trim().to_string())
+                })
+            })
+            .or_else(|| {
+                // Probably we can get the latest version from Maven?
+                Some("1.40.0".to_string())
+            }) {
+            Some(version) => version,
+            None => {
+                zed::set_language_server_installation_status(
+                    language_server_id,
+                    &zed::LanguageServerInstallationStatus::Failed(
+                        "no version specified in settings and no latest release found".to_string(),
+                    ),
+                );
+                return Err(
+                    "no version specified in settings and no latest release found".to_string(),
+                );
+            }
+        };
```

Comment:
Do correct me if I'm wrong, but couldn't this match statement be (off the top of my head) something like `.or_else(|| /* ... */).ok_or("Error message")?` instead?

(Thanks so much for this btw! This has absolutely been needed ^^)

Reply:
@valentinegb 
Yeah, this is some cruft when I tried to use `zed::GithubReleaseOption` to fetch latest versions, but realized they only add tags, not releases.
I simplified it to default to fixed version. -> 65b3bccfffb5e23e06835a9d33456c415401d92f